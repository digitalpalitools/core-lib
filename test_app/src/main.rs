use rusqlite::{Connection, Row, NO_PARAMS};

fn get_row_cells(row: &Row) -> Vec<String> {
    let cells: Vec<String> = row
        .column_names()
        .iter()
        .map(|&cn| {
            let cell: String = match row.get(cn) {
                Ok(c) => c,
                Err(e) => e.to_string(),
            };
            cell
        })
        .collect();

    cells
}

fn exec_sql_core(sql: &str) -> rusqlite::Result<Vec<Vec<Vec<String>>>, rusqlite::Error> {
    let conn = Connection::open("./inflections.db")?;
    let mut result: Vec<Vec<Vec<String>>> = Vec::new();
    for s in sql.split(';').filter(|s| !s.trim().is_empty()) {
        let mut stmt = conn.prepare(&s)?;
        let mut rows = stmt.query(NO_PARAMS)?;

        let mut table: Vec<Vec<String>> = Vec::new();
        while let Some(row) = rows.next()? {
            table.push(get_row_cells(row));
        }
        result.push(table)
    }

    Ok(result)
}

fn exec_sql(sql: &str) -> Result<String, String> {
    let table = exec_sql_core(&sql).map_err(|x| x.to_string())?;
    serde_json::to_string(&table).map_err(|x| x.to_string())
}

fn main() -> Result<(), String> {
    println!("{:?}", pls_core::alphabet::PALI_ALPHABET_ROMAN);
    let x = pls_core::alphabet::PaliAlphabet::AA;
    println!("ā > bh? {:#?}", x > pls_core::alphabet::PaliAlphabet::BH);

    let html = pls_core::inflections::generate_inflection_table(
        "vyābādhetvā",
        |s| Ok(s.to_string()),
        exec_sql,
    )?;
    println!("{:#?}", html);

    std::fs::write("d:/delme/inflections.txt", &html).unwrap();

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    // TODO: Need to be abstracted, granularized and moved to pls_core.
    fn basic_inflection_test() {
        let html = pls_core::inflections::generate_inflection_table(
            "ābādheti",
            |s| Ok(s.to_string()),
            exec_sql,
        );

        let approved_html = include_str!("test_data/basic_inflection_test.approved.txt");

        assert_eq!(html.unwrap(), approved_html);
    }
}
