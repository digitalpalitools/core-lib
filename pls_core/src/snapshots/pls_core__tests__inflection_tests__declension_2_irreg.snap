---
source: test_app/src/main.rs
expression: html

---
<div class="pls-inflection-root">
  <header class="pls-inflection-header">
    <summary class="pls-inflection-summary">kamma 1 &ndash; "kamma nt" (like )</summary>
    </header>
<table class="pls-inflection-table pls-inflection-type-declension">
  <thead>
    <tr>
      <td><span class="pls-inflection-table-title">kamma nt</span></td>
      <td colspan="2"><span class="pls-inflection-col-header">masculine</span></td>
      <td colspan="2"><span class="pls-inflection-col-header">feminine</span></td>
      <td colspan="2"><span class="pls-inflection-col-header">neuter</span></td>
    </tr>
    <tr>
      <td><span class="pls-inflection-row-header">case</span></td>
      <td><span class="pls-inflection-col-header">sg</span></td>
      <td><span class="pls-inflection-col-header">pl</span></td>
      <td><span class="pls-inflection-col-header">sg</span></td>
      <td><span class="pls-inflection-col-header">pl</span></td>
      <td><span class="pls-inflection-col-header">sg</span></td>
      <td><span class="pls-inflection-col-header">pl</span></td>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <p><span class="pls-inflection-row-header">nom</span></p>
      </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        <div class="pls-inflection-inflected-word">kamma</div>
        <div class="pls-inflection-inflected-word">kammaṃ</div>
        </td>
      <td>
        <div class="pls-inflection-inflected-word">kammāni</div>
        <div class="pls-inflection-inflected-word">kammā</div>
        </td>
    </tr>
    <tr>
      <td>
        <p><span class="pls-inflection-row-header">acc</span></p>
      </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        <div class="pls-inflection-inflected-word">kamma</div>
        <div class="pls-inflection-inflected-word">kammaṃ</div>
        </td>
      <td>
        <div class="pls-inflection-inflected-word">kammāni</div>
        <div class="pls-inflection-inflected-word">kamme</div>
        </td>
    </tr>
    <tr>
      <td>
        <p><span class="pls-inflection-row-header">instr</span></p>
      </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        <div class="pls-inflection-inflected-word">kammena</div>
        <div class="pls-inflection-inflected-word">kammunā</div>
        <div class="pls-inflection-inflected-word">kammanā</div>
        </td>
      <td>
        <div class="pls-inflection-inflected-word">kammehi</div>
        </td>
    </tr>
    <tr>
      <td>
        <p><span class="pls-inflection-row-header">dat</span></p>
      </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        <div class="pls-inflection-inflected-word">kammassa</div>
        <div class="pls-inflection-inflected-word">kammuno</div>
        </td>
      <td>
        <div class="pls-inflection-inflected-word">kammānaṃ</div>
        </td>
    </tr>
    <tr>
      <td>
        <p><span class="pls-inflection-row-header">abl</span></p>
      </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        <div class="pls-inflection-inflected-word">kammasmā</div>
        <div class="pls-inflection-inflected-word">kammamhā</div>
        <div class="pls-inflection-inflected-word">kammunā</div>
        </td>
      <td>
        <div class="pls-inflection-inflected-word">kammehi</div>
        </td>
    </tr>
    <tr>
      <td>
        <p><span class="pls-inflection-row-header">gen</span></p>
      </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        <div class="pls-inflection-inflected-word">kammassa</div>
        <div class="pls-inflection-inflected-word">kammuno</div>
        </td>
      <td>
        <div class="pls-inflection-inflected-word">kammānaṃ</div>
        </td>
    </tr>
    <tr>
      <td>
        <p><span class="pls-inflection-row-header">loc</span></p>
      </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        <div class="pls-inflection-inflected-word">kamme</div>
        <div class="pls-inflection-inflected-word">kammasmiṃ</div>
        <div class="pls-inflection-inflected-word">kammamhi</div>
        <div class="pls-inflection-inflected-word">kammani</div>
        </td>
      <td>
        <div class="pls-inflection-inflected-word">kammesu</div>
        </td>
    </tr>
    <tr>
      <td>
        <p><span class="pls-inflection-row-header">voc</span></p>
      </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        <div class="pls-inflection-inflected-word">kamma</div>
        </td>
      <td>
        <div class="pls-inflection-inflected-word">kammāni</div>
        <div class="pls-inflection-inflected-word">kammā</div>
        </td>
    </tr>
    <tr>
      <td><span class="pls-inflection-row-header">in comps</span></td>
      <td colspan="6">
        <div class="pls-inflection-inflected-word">kamma</div>
        </td>
    </tr>
  </tbody>
</table>

  <footer class="pls-inflection-footer">
    <a class="pls-inflection-feedback-link" target="_blank" href="https://docs.google.com/forms/d/e/1FAIpQLSeoxZiqvIWadaLeuXF4f44NCqEn49-B8KNbSvNer5jxgRYdtQ/viewform?entry.267696921=test case&entry.1433863141=v0.1">
      spot a mistake? something missing? fix it here!
    </a>
  </footer>
</div>

